tests/integration/pilot/mcs/discoverability/discoverability_test.go

```diff
@@ -0,0 +1,241 @@
+
+			dest := echos.GetOrFail(ctx, echo.Service(serviceB))
+			for index, c := range ctx.Clusters() {
+				src := echos.GetOrFail(ctx, echo.Service(serviceA).And(echo.InCluster(c)))
+
+				ctx.NewSubTestf("from cluster %d", index).RunParallel(func(ctx framework.TestContext) {
+					// Ensure that all clusters are reached.
+					checkClustersReached(ctx, src, dest, ctx.Clusters())
+				})
+			}
+		})
+}
+
+func TestServiceExportedInOneCluster(t *testing.T) {
+	framework.NewTest(t).
+		Features("traffic.mcs.servicediscovery").
+		Run(func(ctx framework.TestContext) {
+			// Only export service B in the first cluster.
+			cluster0 := ctx.Clusters()[0]
+			createAndCleanupServiceExport(ctx, serviceB, cluster.Clusters{cluster0})
+
+			dest := echos.GetOrFail(ctx, echo.Service(serviceB))
+			for index, c := range ctx.Clusters() {
+				src := echos.GetOrFail(ctx, echo.Service(serviceA).And(echo.InCluster(c)))
+
+				// Since we're exporting only the endpoints in the clusters[0], depending
+				// on where we call service B from, we'll reach a different set of endpoints.
+				// If we're calling from clusters[0], it will be the same as cluster-local
+				// (i.e. we'll only reach endpoints in clusters[0]). From all other clusters,
+				// we should reach endpoints in that cluster AND clusters[0].
+				expectedClusters := cluster.Clusters{cluster0}
+				if c.Name() != cluster0.Name() {
+					expectedClusters = append(expectedClusters, c)
+				}
+
+				ctx.NewSubTestf("from cluster %d", index).RunParallel(func(ctx framework.TestContext) {
+					// Ensure that the expected clusters are reached.
+					checkClustersReached(ctx, src, dest, expectedClusters)
+				})
+			}
+		})
+}
+
+func newServiceExport(service string) *v1alpha1.ServiceExport {
+	return &v1alpha1.ServiceExport{
+		TypeMeta: v12.TypeMeta{
+			Kind:       "ServiceExport",
+			APIVersion: "multicluster.x-k8s.io/v1alpha1",
+		},
+		ObjectMeta: v12.ObjectMeta{
+			Name:      service,
+			Namespace: testNS,
+		},
+	}
+}
+
+func checkClustersReached(ctx framework.TestContext, src, dest echo.Instance, clusters cluster.Clusters) {
+	ctx.Helper()
+	src.CallWithRetryOrFail(ctx, echo.CallOptions{
+		Target:    dest,
+		Count:     100,
+		PortName:  "http",
+		Validator: echo.ExpectReachedClusters(clusters),
+	})
+}
+
+func createAndCleanupServiceExport(ctx framework.TestContext, service string, clusters cluster.Clusters) {
+	ctx.Helper()
+	serviceExport := newServiceExport(service)
+
+	// Create the ServiceExports in each cluster concurrently.
+	g := errgroup.Group{}
+	for _, c := range clusters {
+		c := c
+		g.Go(func() error {
+			_, err := c.MCSApis().MulticlusterV1alpha1().ServiceExports(testNS).Create(context.TODO(),
+				serviceExport, v12.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("failed creating ServiceExport %s/%s in cluster %s: %v",
+					testNS, serviceB, c.Name(), err)
+			}
+
+			// TODO(nmittler): Is there a better way?
```

Comment:
what are you trying to wait for?

Reply:
propagation of the config changes. To wait properly, we would need to look at each envoy to ensure that it has only the expected endpoints.

Reply:
I think the general best practice here is to retry requests until we get the intended behavior. If you want to avoid against random behavior succeeding due to retries, you can use retry.Converge?